### PR TITLE
Refactor spectator class and game method

### PR DESCRIPTION
**Pull request by ramon-bernardo**
### Pull Request Prelude
- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

#### Changes Proposed
- Spectator class from SpectatorVec to Spectators.
- rename addSpectators to insert, accepts a spectator or a vector.

#### Optional Changes Proposed
- getSpectators game method now return a spectators.

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

### CODE REVIEW

File: src/combat.cpp [hunk 1]
```diff
@@ -701,8 +702,7 @@ void Combat::doCombat(Creature* caster, Creature* target) const
 		}
 	} else {
 		if (!params.aggressive || (caster != target && Combat::canDoCombat(caster, target) == RETURNVALUE_NOERROR)) {
-			SpectatorVec spectators;
-			g_game.map.getSpectators(spectators, target->getPosition(), true, true);
+			Spectators spectators = g_game.map.getSpectators(target->getPosition(), true, true);
```
Comment (nekiro) [inline src/combat.cpp:705]: const Spectators& spectators ...

this applies to every single occurrence of similar code.

**Review by nekiro — CHANGES_REQUESTED** [review]

File: src/combat.cpp (hunk 1 above)
Comment (ramon-bernardo) [reply src/combat.cpp]: Thanks!

File: src/condition.cpp [hunk 2]
```diff
@@ -905,12 +905,14 @@ bool ConditionRegeneration::executeCondition(Creature* creature, int32_t interva
 				message.primary.color = TEXTCOLOR_MAYABLUE;
 				player->sendTextMessage(message);
 
-				SpectatorVec spectators;
-				g_game.map.getSpectators(spectators, player->getPosition(), false, true);
+				Spectators& spectators = g_game.map.getSpectators(player->getPosition(), false, true);
```
Comment (nekiro) [inline src/condition.cpp:908]: this cant be reference

File: src/condition.cpp [hunk 3]
```diff
@@ -936,12 +938,14 @@ bool ConditionRegeneration::executeCondition(Creature* creature, int32_t interva
...
```
Comment (nekiro) [inline src/condition.cpp:941]: same as above

File: src/condition.cpp (hunk 2 above)
Comment (ramon-bernardo) [reply src/condition.cpp]: Thanks!

File: src/spectators.h [hunk 4]
```diff
@@ -6,46 +6,61 @@
...
+
+	bool empty() const { return spectators.empty(); }
+
+	void emplace_back(Creature* c) { spectators.emplace_back(c); }
+
+	Iterator begin() { return spectators.begin(); }
+	ConstIterator begin() const { return spectators.begin(); }
+
+	Iterator end() { return spectators.end(); }
+	ConstIterator end() const { return spectators.end(); }
```
Comment (ranisalt) [inline src/spectators.h:60]: Return `decltype(auto)` from these functions and remove the aliases

**Review by ranisalt — CHANGES_REQUESTED** [review]

File: src/spectators.h (hunk 4 above)
Comment (ramon-bernardo) [reply src/spectators.h]: Thanks!

File: src/spectators.h [hunk 5]
```diff
@@ -6,46 +6,55 @@
...
 	{
-		for (Creature* spectator : spectators.vec) {
-			auto it = std::find(vec.begin(), vec.end(), spectator);
-			if (it != end()) {
-				continue;
-			}
-			vec.emplace_back(spectator);
+		auto it = std::find(spectators.begin(), spectators.end(), spectator);
+		if (it == end()) {
+			spectators.emplace_back(spectator);
```
[the remark this answers is not in the document]
Comment (ramon-bernardo) [reply src/spectators.h]: Thanks!

**Pull request closed without merging**